java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java

```diff
@@ -551,7 +576,33 @@ protected KinesisProducer(File inPipe, File outPipe) {
     public int getOutstandingRecordsCount() {
         return futures.size();
     }
-    
+
+    /**
+     * Get the time in seconds for the oldest record currently waiting in kpl to be processed for sending to kinesis
+     * endpoint. The records could either be waiting to be sent to the child process, or have
+     * reached the child process and are being worked on.
+     *
+     * <p>
+     * This is time in seconds from the oldest future submitted from {@link #addUserRecord}
+     * that have not finished. This returns -1 if there no pending records in processing state.
+     *
+     * @return The time in seconds since the oldest record is pending to be sent to kinesis endpoint. Returns -1 if
+     * there are no records in processing state.
+     */
+    @Override
+    public long getOldestRecordTimeInSeconds() {
+        long oldestTime = Long.MAX_VALUE;
+        for (final Map.Entry<Long, SettableFutureTracking> entry : futures.entrySet()) {
+            if (entry.getValue().getTimestamp().getEpochSecond() <  oldestTime) {
+                oldestTime = entry.getValue().getTimestamp().getEpochSecond();
+            }
+        }
+        if(oldestTime == Long.MAX_VALUE) {
+            return -1;
+        }
+        return Instant.now().getEpochSecond() - oldestTime;
+    }
```

Comment:
Can we use a minheap data structure to store and retrieve the oldest future in O(1) ?